### PR TITLE
add Pairwise distance

### PR DESCRIPTION
**Pull request by edgarriba**

### CODE REVIEW

**Comment by edgarriba** [comment]
@apaszke could you give me some instructions to add tests ?

**Comment by szagoruyko** [comment]
@edgarriba how is it different from `torch.dist` ? http://pytorch.org/docs/torch.html?highlight=dist#torch.dist

**Comment by edgarriba** [comment]
@szagoruyko `torch.dist` returns a scalar instead of a distances vector

**Comment by szagoruyko** [comment]
@edgarriba ah I see, thanks. I think it better fits in `torch.nn.functional`, there are no parameters for a Module

**Comment by fmassa** [comment]
I agree with @szagoruyko, it would be good to have it in `torch.nn.functional`. But do we want to have full parity between `torch.nn.functional` and `torch.nn` (meaning that we will always have to wrap the function in a `nn.Module` as well)?

File: torch/nn/modules/distance.py [hunk 1]
```diff
@@ -1,6 +1,25 @@
...
+    """
+
+    def __init__(self, p=2):
+        super(PairwiseDistance, self).__init__()
+        assert p % 1 == 0
+        self.norm = p
+
+    def forward(self, x1, x2):
+        assert x1.size() == x2.size()
+        eps = 1e-4 / x1.size(1)
```
Comment (fmassa) [inline torch/nn/modules/distance.py]: I think it might be better to pass `eps` as a parameter in the constructor, and maybe something smaller and independent on the feat dimension like `eps=1e-6` for example).

File: torch/nn/modules/distance.py [hunk 2]
```diff
@@ -1,6 +1,25 @@
...
+        diff = torch.abs(x1 - x2)
+        out = torch.pow(diff, self.norm).sum(dim=1)
+        return torch.pow(out + eps, 1. / self.norm)
```
Comment (fmassa) [inline torch/nn/modules/distance.py]: Not sure about it in here, but maybe this is not a good expression for the `1`-norm when `x1 == x2`.
In torch, the gradients in such case are randomly assigned to `-1` or `1`, [as can be seen here](https://github.com/torch/nn/blob/master/PairwiseDistance.lua#L37).
In here, the gradients are always assigned to `1` when `x1 == x2` (because `eps` is positive).

Not sure how it can affect things though.

File: torch/nn/modules/distance.py (hunk 1 above)
Comment (edgarriba) [inline torch/nn/modules/distance.py]: this was the first solution that came out from the slack discussion. Let me try with eps 1e-6 and check that my network still converges

Comment (fmassa) [inline torch/nn/modules/distance.py]: I think it should converge, as it is the value that is added in the [`torch` version](https://github.com/torch/nn/blob/master/PairwiseDistance.lua#L64)

File: torch/nn/modules/distance.py (hunk 2 above)
Comment (edgarriba) [inline torch/nn/modules/distance.py]: nor a strong argument but since is random, case with gradient `1` can happen, so I don't see the problem

Comment (apaszke) [inline torch/nn/modules/distance.py]: I think it's better to have a special case for 1 norm. Also, why does it support only integer `p`s? It should work for floats too, right?

**Comment by edgarriba** [comment]
ping @apaszke why is not passing tests?

**Comment by fmassa** [comment]
@edgarriba this is probably due to https://github.com/pytorch/pytorch/pull/952 and is not an issue here,

**Comment by edgarriba** [comment]
@fmassa ops, I'll rebase. BTW, not sure how to handle 1norm since the difference is in the backward pass and we are assuming to use autograd. Besides, should we rename pairwise distance to pnorm?

**Comment by fmassa** [comment]
The first thing that comes to my mind is to define a `Function` with a hand-coded backward for the cases that we are targeting. But I don't like that solution that much, and maybe it would be better to address those instability issues in the underlying implementations of the necessary functions (`div`, `pow` etc).

File: test/test_nn.py [hunk 3]
```diff
@@ -2349,6 +2349,10 @@ def add_test(test):
         input_size=(1, 2, 4, 4),
         desc='scale'
     ),
+    dict(
+        constructor=lambda: nn.PairwiseDistance(),
```
Comment (apaszke) [inline test/test_nn.py]: Replace that with `module_name='PairwiseDistance'`

File: torch/nn/functional.py [hunk 4]
```diff
@@ -565,3 +565,19 @@ def pad(input, pad, mode='constant', value=0):
...
+
+    Args
+        input (Variable): Tensor with size (2, batch, dim)
+        p (int) default=2: The norm degree
+    """
+    assert len(input) == 2, "Input needs to be of size (2, batch, dim)"
+    x1, x2 = input
+    diff = torch.abs(x1 - x2)
+    out = torch.pow(diff, p).sum(dim=1)
+    return torch.pow(out + 1e-6, 1. / p)
```
Comment (apaszke) [inline torch/nn/functional.py]: Not sure how common it is to use `p < 1`, but for these values its the first `pow` that will be numerically unstable and the eps needs to be added there instead.

File: torch/nn/modules/distance.py [hunk 5]
```diff
@@ -1,6 +1,29 @@
...
+
+
+class PairwiseDistance(Module):
+    """
+    Computes the batchwise pairwise distance between vectors v1,v2:
+
+        .. math ::
+            \Vert x \Vert _p := \left( \sum_{i=1}^n  \vert x_i \vert ^ p \right) ^ {1/p}
+
+        Inputs: x
```
Comment (apaszke) [inline torch/nn/modules/distance.py]: We're using a different convention for inputs and outputs (`Arguments` or `Args` and `Returns`). Please make it consistent with documentation of other modules.

**Review by apaszke — COMMENTED** [review]
What's the problem with 1-norm? It should work in that case too, right? (both pows should be no-ops)

File: torch/nn/functional.py [hunk 6]
```diff
@@ -565,3 +565,31 @@ def pad(input, pad, mode='constant', value=0):
...
+
+        .. math ::
+            \Vert x \Vert _p := \left( \sum_{i=1}^n  \vert x_i \vert ^ p \right) ^ {1/p}
+
+        Args:
+            x (Tensor): input tensor containing the two input batches
+            p (real): the norm degree. Default: 2
+
+        Shape:
+            - Input: :math:`(2, N, D)` where `D = vector dimension`
```
Comment (apaszke) [inline torch/nn/functional.py]: Actually there's no point in requiring a single input to the module and the function. You could easily compute outputs on two batches with different models, and ignore the cat before feeding them into the pairwise distance. Can you change the signature to `pairwise_distance(x1, x2, p=2, eps=1e-6)` (and do the same with the module)? It's also helpful to be able to change the eps.

File: torch/nn/functional.py [hunk 7]
```diff
@@ -565,3 +565,31 @@ def pad(input, pad, mode='constant', value=0):
...
+            - Output: :math:`(N, 1)
+
+        >>> pdist = nn.PairwiseDistance(2)
```
Comment (apaszke) [inline torch/nn/functional.py]: You should use the function in the functional docs

File: torch/nn/functional.py (hunk 6 above)
Comment (edgarriba) [inline torch/nn/functional.py]: sure no problem

**Comment by fmassa** [comment]
This looks good @edgarriba , thanks a lot! (and sorry for asking so many things in this PR)

**Comment by apaszke** [comment]
You forgot to modify the tests 😕

File: torch/nn/functional.py [hunk 8]
```diff
@@ -565,3 +565,32 @@ def pad(input, pad, mode='constant', value=0):
...
+            - Input: :math:`(N, D)` where `D = vector dimension`
+            - Output: :math:`(N, 1)
+
+        >>> input1 = autograd.Variable(torch.randn(100, 128))
+        >>> input2 = autograd.Variable(torch.randn(100, 128))
+        >>> output = F.pairwise_distance(input1, input2, p=2)
+        >>> output.backward()
+    """
+    assert x1.size() == x2.size(), "Input sizes must be equal."
+    assert x1.dim() == 2, "Input rank must be equal to two."
```
Comment (apaszke) [inline torch/nn/functional.py]: Well, number of dimensions isn't really a matrix rank 😕 I'd just say that it should be a 2D tensor.

**Comment by edgarriba** [comment]
@apaszke any update for test arguments?

**Comment by apaszke** [comment]
@edgarriba no, sorry for that. I'll try to add them tomorrow.

**Comment by apaszke** [comment]
Ok, I've pushed a commit that enables testing of multiarg modules into a `pairwise-distance` branch in this repo. Can you include it in this branch (feel free to squash it) and fix the tests?

**Comment by soumith** [comment]
@edgarriba the commit history now looks screwed up.

**Comment by edgarriba** [comment]
are you squashing during the merge?

**Comment by soumith** [comment]
yes, i'll squash no worries. Now the commits look better :)

**Comment by edgarriba** [comment]
nice thx!

**Comment by edgarriba** [comment]
@apaszke any hint?

**Comment by apaszke** [comment]
The tests depend on the input being a single Variable in more places than I originally thought. I need more time to fix that, or you can just temporarily add a regular test method to the `TestNN` class. You can use `autograd.gradcheck`.

File: test/test_nn.py [hunk 9]
```diff
@@ -2035,6 +2040,24 @@ def test_orthogonal(self):
                                          torch.eye(rows) * gain ** 2, prec=1e-6)
 
 
+class unpack_args(object):
```
Comment (apaszke) [inline test/test_nn.py]: this is now unnecessary, can you please remove it?

**Review by apaszke — APPROVED** [review]

File: test/test_nn.py [hunk 10]
```diff
@@ -1763,6 +1763,11 @@ def test_batchnorm_eval(self):
             self.assertEqual(res1, res2)
             self.assertEqual(grad1, grad2)
 
+    def test_pairwise_distance(self):
+        input1 = Variable(torch.randn(4, 4), requires_grad=True)
+        input2 = Variable(torch.randn(4, 4), requires_grad=True)
+        gradcheck(lambda x, y: F.pairwise_distance(x, y), (input1, input2))
```
Comment (fmassa) [inline test/test_nn.py]: You need to assert that the return value of `gradcheck` is `True`

Comment (edgarriba) [inline test/test_nn.py]: So also is needed here?
https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L675

Comment (szagoruyko) [inline test/test_nn.py]: @edgarriba yes :P forgot to push a fix for this

Comment (edgarriba) [inline test/test_nn.py]: I'll open a separated PR for this to not​ mix things

**Review by apaszke — APPROVED** [review]
